internal/data.go

```diff
@@ -0,0 +1,153 @@
+// Copyright 2022 Tigris Data, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package internal
+
+import (
+	"bytes"
+	"time"
+
+	api "github.com/tigrisdata/tigrisdb/api/server/v1"
+	"github.com/ugorji/go/codec"
+	"google.golang.org/grpc/codes"
+)
+
+var (
+	bh codec.BincHandle
+)
+
+// DataType is to define the different data types for the data stored in the storage engine.
+type DataType byte
+
+// Note: Do not change the order. Order is important because encoder is adding the type as the first byte.
+const (
+	Unknown DataType = iota
+	TableDataType
+	CDCDataType
+	DictEncDataType
+	RawDataType
+)
+
+// Data interface acts as a value wrapper for the raw bytes.
+type Data interface {
+	GetDataType() DataType
+}
+
+func NewTimestamp() *Timestamp {
+	return &Timestamp{
+		Nano: time.Now().UnixNano(),
+	}
+}
+
+// NewTableData returns a table data type by setting the ts to the current value.
+func NewTableData(data []byte) *TableData {
+	return &TableData{
+		Ts:       NewTimestamp(),
+		UserData: data,
+	}
+}
+
+func (x *TableData) GetDataType() DataType {
+	return TableDataType
+}
+
+func (x *CDCData) GetDataType() DataType {
+	return CDCDataType
+}
+
+func NewDictEncData(encValue uint32) *DictEncData {
+	return &DictEncData{
+		Ts:       NewTimestamp(),
+		EncValue: encValue,
+	}
+}
+
+func (x *DictEncData) GetDataType() DataType {
+	return DictEncDataType
+}
+
+func NewRawData(data []byte) *RawData {
+	return &RawData{
+		Ts:   NewTimestamp(),
+		Data: data,
+	}
+}
+
+func (x *RawData) GetDataType() DataType {
+	return RawDataType
+}
+
+func Encode(data Data) ([]byte, error) {
+	var buf bytes.Buffer
+	err := buf.WriteByte(byte(data.GetDataType()))
+	if err != nil {
+		return nil, err
+	}
+	enc := codec.NewEncoder(&buf, &bh)
+
+	switch data.GetDataType() {
+	case TableDataType:
+		if err := enc.Encode(data.(*TableData)); err != nil {
```

Comment:
If there is some bug, this may panic.
Can we `switch t := xxx.(type)` instead?

Reply:
sure